### PR TITLE
[refactor] Add a powerful `repr()` function for debugging

### PR DESCRIPTION
**Pull request by vkarak**
This function replaces the older `debug.repr()` and uses `util.ppretty()` for nicely printing the object attributes. `util.ppretty()` is also extended in order to recursively call the new `repr()` function. Objects are printed only once by this new `repr()` avoiding infinite recursions.

Finally, we use that function to print the whole ReFrame runtime state in case #1369 appears.

This function does not take into account the `__slots__` for the moment. In cases, where an object does not have a `__dict__` attribute, the builtin `repr()` will be used.

An example output for a list of objects with self references is the following (see also the corresponding unit test):

```
[
    _Y({
        'x': 1,
        'y': 'foo',
        'z': _Y(...)@0x102463550,
        'a': None
    })@0x102463550,
    _Y({
        'x': 2,
        'y': 'foo',
        'z': _Y(...)@0x102463510,
        'a': _X({
            '_a': False
        })@0x102463290
    })@0x102463510
]
```

### CODE REVIEW

File: unittests/test_policies.py [hunk 1]
```diff
@@ -524,6 +525,8 @@ def test_kbd_interrupt_in_wait_with_limited_concurrency(
             KeyboardInterruptCheck(), SleepCheck(10),
             SleepCheck(10), SleepCheck(10)
         ]))
+        # FIXME: Dump everything in case Github #1369 appears
+        print(util.repr(runner))
```
Comment (vkarak) [inline unittests/test_policies.py:529]: Unindent this and try running this unit test with `-s` to check what this innocent line prints ;-)

**Comment by vkarak** [comment]
Check comment in issue #1369 as a proof that this PR works wonderfully :-) It served its purpose already!

File: reframe/utility/__init__.py [hunk 2]
```diff
@@ -164,24 +170,55 @@ def ppretty(value, htchar=' ', lfchar='\n', indent=4, basic_offset=0):
...
+        if hasattr(obj, '__dict__'):
+            objects.add(addr)
+
+        return func(obj, **kwargs)
+
+    return _repr
+
+
+@_tracked_repr
+def repr(obj, *, htchar=' ', lfchar='\n', indent=4, basic_offset=0):
```
Comment (vkarak) [inline reframe/utility/__init__.py:211]: Actually, here I don't need all the other options... I could update the way I'm trying to call `repr()` in `ppretty()`.

File: reframe/utility/__init__.py [hunk 3]
```diff
@@ -135,26 +137,31 @@ def ppretty(value, htchar=' ', lfchar='\n', indent=4, basic_offset=0):
     :arg indent: Number of htchar characters for every indentation level.
     :arg basic_offset: Basic offset for the representation, any additional
         indentation space is added to the ``basic_offset``.
+    :arg repr: The :func:`repr` to use for printing values. This function may
+        accept also all the arguments of :func:`ppretty` except the ``repr``.
+
     :returns: a formatted string of the ``value``.
     '''
 
+    print(repr)
```
Comment (teojgo) [inline reframe/utility/__init__.py:146]: Why do you need this print?

Comment (vkarak) [reply reframe/utility/__init__.py]: No, it's stale...

**Pull request merged**